### PR TITLE
Issue 179 kb metadata

### LINKED ISSUE #179: KB metadata information service
An API service with metadata information pertaining to the KB build is needed, containing at least the following:

- Build date (currently part of `/kb/annotation_summary`) and/or other KB version identification (see also #17)
- Versions of ontologies used in KB build
- Identification of software and their versions used in building the KB

### PR DESCRIPTION
**Pull request by Shalsh23**
Fixes #179

### CODE REVIEW

File: src/main/scala/org/phenoscape/kb/KB.scala [hunk 1]
```diff
@@ -189,6 +190,41 @@ OPTIONAL {
...
+      ?ont $rdfType owl:Ontology .
+  	  ?ont owl:versionIRI ?version
+    }
+  """
+
+    App.executeSPARQLQueryString(query.text, qs => IRI.create(qs.getResource("version").getURI))
+  }
+}
+
+case class KBMetadata(built: Instant, ontologies: Seq[IRI]) {
```
Comment (balhoff) [inline src/main/scala/org/phenoscape/kb/KB.scala:217]: Extend JSONResultItem. Then you don't need the customer marshaller for this.

File: src/main/scala/org/phenoscape/kb/KB.scala [hunk 2]
```diff
@@ -189,6 +190,41 @@ OPTIONAL {
...
+
+  def toJSON: JsObject = Map(
+    "build_time" -> built.toString.toJson,
+    "kb ontologies" -> ontologies.map(_.toString.toJson).toJson
```
Comment (balhoff) [inline src/main/scala/org/phenoscape/kb/KB.scala:221]: Just a personal preference: I like to use underscores instead of spaces in keys. Just call this "ontologies".

File: src/main/scala/org/phenoscape/kb/KB.scala (hunk 1 above)
Comment (balhoff) [reply src/main/scala/org/phenoscape/kb/KB.scala:217]: Also, each ontology should be an object containing its IRI and also its version IRI.

File: src/main/scala/org/phenoscape/kb/KB.scala [hunk 3]
```diff
@@ -189,8 +190,44 @@ OPTIONAL {
     App.executeSPARQLQueryString(query.text, res => Instant.parse(res.getLiteral("date").getLexicalForm)).map(_.head)
   }
 
+  def getKBMetadata = {
```
Comment (balhoff) [inline src/main/scala/org/phenoscape/kb/KB.scala:193]: Methods should have type ascriptions.

File: src/main/scala/org/phenoscape/kb/KB.scala [hunk 4]
```diff
@@ -189,8 +190,44 @@ OPTIONAL {
...
+    val builtFut = buildDate
+    val ontologiesFut = kbOntologies
+    for {
+      built <- builtFut
+      ontologies <- ontologiesFut
+    } yield KBMetadata(built, ontologies)
+  }
+
+  def kbOntologies = {
```
Comment (balhoff) [inline src/main/scala/org/phenoscape/kb/KB.scala:202]: Methods should have type ascriptions.

File: src/main/scala/org/phenoscape/kb/KB.scala [hunk 5]
```diff
@@ -189,8 +190,44 @@ OPTIONAL {
...
+  }
 }
 
+case class KBMetadata(built: Instant, ontologies: Seq[(IRI, IRI)]) extends JSONResultItem {
+
+  def toJSON: JsObject = Map(
+    "build_time" -> built.toString.toJson,
+    "ontologies" -> ontologies.map{
+      case(ont, version) => JsObject(
+        "ontololgy" -> ont.toString.toJson,
```
Comment (balhoff) [inline src/main/scala/org/phenoscape/kb/KB.scala:223]: Typo

Comment (balhoff) [reply src/main/scala/org/phenoscape/kb/KB.scala:223]: To go along with our JSON-LD-ish convention we use elsewhere, use key `@id` instead of `ontology`.

File: src/main/scala/org/phenoscape/kb/Main.scala [hunk 6]
```diff
@@ -121,6 +121,12 @@ object Main extends HttpApp with App {
           pathSingleSlash {
             redirect(Uri("http://kb.phenoscape.org/apidocs/"), StatusCodes.SeeOther)
           } ~ pathPrefix("kb") {
+            path("metadata") {
+              complete {
+                import org.phenoscape.kb.JSONResultItem.JSONResultItemsMarshaller
```
Comment (balhoff) [inline src/main/scala/org/phenoscape/kb/Main.scala:126]: Why is this needed?

File: src/main/scala/org/phenoscape/kb/KB.scala (hunk 5 above)
Comment (Shalsh23) [reply src/main/scala/org/phenoscape/kb/KB.scala]: > Typo

Where?

Comment (balhoff) [reply src/main/scala/org/phenoscape/kb/KB.scala]: "ontololgy". But as I said I prefer not to have this key anyway.

File: swagger/swagger.yaml [hunk 7]
```diff
@@ -23,6 +23,26 @@ tags:
...
+        200:
+          description: KB metadata
+          examples:
+            application/json:
+              build_time: '2020-04-08T20:37:21.892Z'
+              ontologies:
+                - "@id": http://purl.obolibrary.org/obo/pato.owl
+                  version: http://purl.obolibrary.org/obo/pato/releases/2020-03-24/pato.owl
+                - "@id": http://purl.obolibrary.org/obo/ro.owl
+                  version: http://purl.obolibrary.org/obo/ro/releases/2020-02-26/ro.owl
```
Comment (Shalsh23) [inline swagger/swagger.yaml:45]: @balhoff Need your feedback here. There is some issue which is eluding me.

Comment (balhoff) [reply swagger/swagger.yaml:45]: What do you mean? Looks good.

File: swagger/swagger.yaml [hunk 8]
```diff
@@ -2577,3 +2605,6 @@ definitions:
         type: array
         items:
           $ref: '#/definitions/TermWithLabel'
+# Added by API Auto Mocking Plugin
+host: virtserver.swaggerhub.com
+basePath: /Shalsh23/phenoscape-kb-webservices/1.0.0
```
Comment (balhoff) [inline swagger/swagger.yaml:2610]: Please remove.

File: src/main/scala/org/phenoscape/kb/KB.scala [hunk 9]
```diff
@@ -189,8 +190,44 @@ OPTIONAL {
...
+  def getKBMetadata: Future[KBMetadata]= {
+    val builtFut = buildDate
+    val ontologiesFut = kbOntologies
+    for {
+      built <- builtFut
+      ontologies <- ontologiesFut
+    } yield KBMetadata(built, ontologies)
+  }
+
+  def kbOntologies: Future[Seq[(IRI, IRI)]] = {
```
Comment (balhoff) [inline src/main/scala/org/phenoscape/kb/KB.scala:202]: Use `Set` instead of `Seq` here.

File: src/main/scala/org/phenoscape/kb/KB.scala [hunk 10]
```diff
@@ -189,8 +190,44 @@ OPTIONAL {
...
+      ?ont $rdfType owl:Ontology .
+  	  ?ont owl:versionIRI ?version
+    }
+  """
+
+    App.executeSPARQLQueryString(query.text, qs => (IRI.create(qs.getResource("ont").getURI), IRI.create(qs.getResource("version").getURI)))
+  }
 }
 
+case class KBMetadata(built: Instant, ontologies: Seq[(IRI, IRI)]) extends JSONResultItem {
```
Comment (balhoff) [inline src/main/scala/org/phenoscape/kb/KB.scala:217]: Use `Set` instead of `Seq` here.

File: src/main/scala/org/phenoscape/kb/KB.scala [hunk 11]
```diff
@@ -189,8 +190,44 @@ OPTIONAL {
...
+
+  def toJSON: JsObject = Map(
+    "build_time" -> built.toString.toJson,
+    "ontologies" -> ontologies.map{
```
Comment (balhoff) [inline src/main/scala/org/phenoscape/kb/KB.scala:221]: Even though I am suggesting to use `Set` in the constructor, I would then convert to Seq here and sort somehow before converting to JSON, to ensure consistent output.

File: swagger/swagger.yaml [hunk 12]
```diff
@@ -9,8 +9,8 @@ info:
   license:
     name: MIT
     url: https://opensource.org/licenses/MIT
-host: kb.phenoscape.org
-basePath: /api
+# host: kb.phenoscape.org
+# basePath: /api
```
Comment (balhoff) [inline swagger/swagger.yaml:13]: Is this on purpose?

Comment (Shalsh23) [reply swagger/swagger.yaml:13]: No, this was probably already there on swagger hub where I pasted this from. But I'll remove it.

File: src/main/scala/org/phenoscape/kb/KB.scala [hunk 13]
```diff
@@ -189,8 +190,44 @@ OPTIONAL {
...
+
+    App.executeSPARQLQueryString(query.text, qs => (IRI.create(qs.getResource("ont").getURI), IRI.create(qs.getResource("version").getURI))).map(_.toSet)
+  }
 }
 
+case class KBMetadata(built: Instant, ontologies: Set[(IRI, IRI)]) extends JSONResultItem {
+
+  def toJSON: JsObject = Map(
+    "build_time" -> built.toString.toJson,
+    "ontologies" -> ontologies.toSeq.sortBy(_.toString()).map{
```
Comment (balhoff) [inline src/main/scala/org/phenoscape/kb/KB.scala:221]: Remove parentheses on `toString` call (using parentheses on nullary methods is Scala convention for side-effecting call). Need a space within `map{`.

File: swagger/swagger.yaml [hunk 14]
```diff
@@ -9,8 +9,7 @@ info:
   license:
     name: MIT
     url: https://opensource.org/licenses/MIT
-host: kb.phenoscape.org
-basePath: /api
```
Comment (balhoff) [inline swagger/swagger.yaml:13]: This is still being deleted. Any reason not to put it back?

File: src/main/scala/org/phenoscape/kb/KB.scala (hunk 13 above)
Comment (Shalsh23) [reply src/main/scala/org/phenoscape/kb/KB.scala:221]: Why do we need space after `map`? Is there some place I can read about such conventions?

Comment (balhoff) [reply src/main/scala/org/phenoscape/kb/KB.scala]: Just for consistency (I can't explain why it looks better; I think it does). The default IntelliJ formatter should do this. We could look into integrating [scalafmt](https://scalameta.org/scalafmt/) to separate this from IntelliJ.
